### PR TITLE
WIP: initialize mklfft

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
My attempt at #1916

Please check if the general approach seems okay.
Basically, there is a function in mne.utils, `_get_mkl_fft`, that takes as arguments a string and `n_jobs`. If `n_jobs` isn't CUDA, it tries to import the function denoted by the string from `mklfft.fftpack`. It will also try to import mkl from mkl-service and set the number of threads to n_jobs.
Otherwise, it returns the corresponding function from `scipy.fftpack`.

### CODE REVIEW

File: mne/filter.py [hunk 1]
```diff
@@ -11,7 +16,7 @@
 from .parallel import parallel_func, check_n_jobs
 from .cuda import (setup_cuda_fft_multiply_repeated, fft_multiply_repeated,
                    setup_cuda_fft_resample, fft_resample, _smart_pad)
-from .utils import logger, verbose, sum_squared, check_version
+from .utils import logger, verbose, sum_squared, check_version, _get_mkl_fft
```
Comment (jona-sassenhagen) [inline mne/filter.py]: So in mne.utils, I'd have something like

```
def fft(*args, **kwargs):
    try:
        from mklfft.fftpack import fft
    except ImportError:
        from scipy.fftpack import fft
    return fft(*args, **kwargs)
```

... or if not, can you point me to some code that does what you mean?

I tried modeling this after how ica and xdawn handle fast_dot.

Comment (agramfort) [inline mne/filter.py]: just add to the utils.py file

```
try:
    from mklfft.fftpack import fft
except ImportError:
    from scipy.fftpack import fft
```

it should do the trick

**Comment by jona-sassenhagen** [comment]
Like this? (error is pep)

**Comment by agramfort** [comment]
so far so good.

let us when you have benchmarks.

**Comment by jona-sassenhagen** [comment]
On my i5 iMac, it's actually slower to use mklfft with 4 or 8 threads. I can't currently test it on our actually powerful computers because somebody is using them to run MATLAB ...

**Comment by jasmainak** [comment]
Thanks for taking a stab at this @jona-sassenhagen . Something looks weird with your commit history. You can fix it by rebase

**Comment by jona-sassenhagen** [comment]
Thanks @jasmainak ... not sure it's worth it though, seeing as I don't see improvements. Is anyone else getting speed improvements? I can imagine it might make a difference for systems with many cores, but little RAM, but I don't have such a machine available.

File: mne/utils.py [hunk 2]
```diff
@@ -39,6 +39,23 @@
...
+        warnings.warn("`mklfft` found, but `mkl` could not be loaded. Not able"
+                      " to set the number of threads via mkl.set_num_threads."
+                      " Note that single-threaded `mklfft` is not expected to"
+                      " be faster than Scipy's regular FFT.")
+    logger.info("Using `mklfft` for FFT.")
+except ImportError:
+    from scipy.fftpack import fft, ifft, fftn, ifftn
+
+for f in (fft, ifft, fftn, ifftn):
+    assert f  # to avoid pyflakes complaining about unused imports
```
Comment (dengemann) [inline mne/utils.py]: is this necessary? I think we can live with a small exception here.

**Comment by jasmainak** [comment]
Last I remember when I tried it, I wasn't getting any improvements ...

File: mne/utils.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/utils.py]: better just to use `# noqa` at the end of the offending lines

**Comment by jona-sassenhagen** [comment]
In that thread, Alex said it was due to multithreading being required. But I somehow don't see improvements even when really hammering all of our cores with `mklfft` either. I don't know.

**Comment by jasmainak** [comment]
@jona-sassenhagen : can you clarify how I should read your benchmarks? What is real, user and sys?

**Comment by jona-sassenhagen** [comment]
That's just with standard Unix time - so "real" is the actual time it takes to run the script, and I guess "user" is a measure of CPU time or something.

**Comment by jasmainak** [comment]
Could it be that there aren't many fft operations to make a real difference? Maybe you need to benchmark on something where there are many fft operations ...

**Comment by jona-sassenhagen** [comment]
I'd think filtering raws with `method=fft` is one of the most fft heavy task we have, or is there a better test?

**Comment by jasmainak** [comment]
did you try TF decomposition on epochs? that was real slow. It might be worthwhile to profile and figure out if at least the `fft` operations are faster.

**Comment by jona-sassenhagen** [comment]
Which method (multitaper etc) ... would benefit the most you think?

**Comment by jasmainak** [comment]
No idea ... I remember having tried `tfr_morlet` with `use_fft=True` and that was pretty slow. Also, for profiling: https://github.com/rkern/line_profiler in case you don't know it already

**Comment by jasmainak** [comment]
By pretty slow, I mean it could take 10 minutes to run on my computer ...

**Comment by jona-sassenhagen** [comment]
@jasmainak that's sadly not disproportionally slow for TF analyses I fear. I've had TF decompositions take days to run for a full experiment and low frequencies (= long windows) on EEGLAB.
If you really want speed, CUDA is the way to go.

**Comment by jasmainak** [comment]
@agramfort thoughts?

**Comment by agramfort** [comment]
can you share a bench script?

**Comment by jasmainak** [comment]
@jona-sassenhagen would you have time for this in the coming days? If not, I can take a look once the EEGLAB .set reader is merged.

**Comment by jona-sassenhagen** [comment]
I'm a bit demotivated due to the lack of results ... feel free to take over.

**Comment by jona-sassenhagen** [comment]
Closing this for now as I just opened two other PRs, @Eric89GXL opened his issue on the topic, @jasmainak indicated he might want to take over, and I'm not making progress.

**Pull request closed without merging**